src/generator.cs

```diff
@@ -6241,19 +6272,35 @@ public void Generate (Type type)
 								library_name = CoreServicesMap;
 								break;
 							}
-						} 
+						} else {
+							// we get something in LibraryName from FieldAttribute so we asume
+							// it is a path to a library, so we save the path and change library name
+							// to a valid identifier if needed
+							library_path = library_name;
+							if (library_name.Contains ("/")) {
+								var lastSlashIdx = library_name.LastIndexOf ("/");
+								library_name = library_name.Substring (lastSlashIdx + 1);
+							}
+							if (library_name.Contains ("."))
+								library_name = library_name.Replace (".", string.Empty);
+						}
+					} else if (BindThirdPartyLibrary) {
+						// User should provide a LibraryName
+						throw new BindingException (1042, true, $"Missing '[Field (LibraryName=value)]' for {field_pi.Name} (e.g.\"__Internal\")");
 					} else {
 						library_name = type.Namespace;
 						// note: not every binding namespace will start with ns.Prefix (e.g. MonoTouch.)
-						if (!String.IsNullOrEmpty (ns.Prefix) && library_name.StartsWith (ns.Prefix))
+						if (!String.IsNullOrEmpty (ns.Prefix) && library_name.StartsWith (ns.Prefix)) {
 							library_name = library_name.Substring (ns.Prefix.Length + 1);
+							library_name = library_name.Replace (".", string.Empty); // Remove dots from namespaces
+						}
 					}
 
-					if (!libraries.Contains (library_name)) {
-						libraries.Add (library_name);
+					if (!libraries.Any (l => l.Item1 == library_name)) {
+						libraries.Add (new Tuple<string, string> (library_name, library_path));
```

Comment:
Read my comment about maybe using a dictionary to simplify this and make it faster.